lib/shoulda/matchers/active_model/validate_length_of_matcher.rb

```diff
@@ -487,13 +490,29 @@ def disallows_length_of?(length, message)
         end
 
         def value_of_length(length)
-          (array_column? ? ['x'] : 'x') * length
+          if array_column?
+            ['x'] * length
+          elsif association?
+            Array.new(length) { association.klass.new }
+          else
+            'x' * length
+          end
```

Comment:
My initial idea was to multiply by `length` the return of the condition, like the example bellow.
```ruby
if array_column?
  ['x']
elsif association?
  [association.klass.new]
else
  'x'
end * length
```

However, this causes the array of the associated object to have the same `object_id`, and Rails prevents the same object from being associated twice check on [collection_association](https://github.com/rails/rails/blob/48243d42aad9675e26408aaa561e677dac91db04/activerecord/lib/active_record/associations/collection_association.rb#L416-L426) class on active record gem for more context. That's why I needed to change this logic, so we can have different `object_ids` on the array of associations.

https://github.com/rails/rails/blob/48243d42aad9675e26408aaa561e677dac91db04/activerecord/lib/active_record/associations/collection_association.rb#L416-L426